src/orchestrator/orchestrator.rs

```diff
@@ -0,0 +1,191 @@
+        for i in 0..actual_number_of_workers {
+            let (entries_tx, mut entries_rx) = mpsc::channel(32);
+            let (tree_tx, tree_rx) = mpsc::channel(32);
+            let executor = self.executor_spawner.spawn_executor().await;
+            result_collectors.push((i, tree_rx));
+
+            // 2. Executor
+            //
+            // Spawn executors that process entries with Worker.
+            //
+            // - Receives 'entries' from [entries_rx] channel.
+            // - Processes 'entries' to build a tree (done by worker).
+            // - Sends the resulting 'tree' back via [tree_tx] channel.
+            //
+            executors.push(tokio::spawn(async move {
+                while let Some(task) = entries_rx.recv().await {
+                    let processed_task = executor
+                        .generate_tree::<N_ASSETS, N_BYTES>(task)
+                        .await
+                        .unwrap();
+                    if tree_tx.send(processed_task).await.is_err() {
+                        break;
+                    }
+                }
+            }));
+
+            // 1. Distributing Tasks
+            //
+            // Spawn a distribution thread that distributes entries to executors
+            //
+            // - Loads CSV file from [csv_file_path].
+            // - Parses CSV file into 'entries'.
+            // - Sends 'entries' to executors via [entries_tx] channel.
+            //
+            let (start, end) = self.calculate_task_range(i, number_executor);
+            let entry_csvs_slice = self.entry_csvs[start..end].to_vec(); // Clone only the necessary slice
+
+            tokio::spawn(async move {
+                for file_path in entry_csvs_slice.iter() {
+                    let entries = entry_parser::<_, N_ASSETS, N_BYTES>(file_path).unwrap();
+                    if entries_tx.send(entries).await.is_err() {
+                        break;
+                    }
+                }
+            });
+        }
+
+        // 3. Collectoing Results
+        //
+        // Collect `tree` results from executors
+        //
+        //  - Receives processed 'tree' from [tree_rx] channel.
+        //  - Collects all 'tree' results into 'worker_results'.
+        //  - Aggregates 'worker_results' into 'aggregated_tree_results'.
+        //
+        let mut all_tree_responses = Vec::new();
+        for (index, mut result_rx) in result_collectors {
+            let executor_results = tokio::spawn(async move {
+                let mut trees = Vec::new();
+                while let Some(result) = result_rx.recv().await {
+                    trees.push(result);
+                }
+                (index, trees)
+            });
+            all_tree_responses.push(executor_results);
+        }
+
+        // Wait for all workers to finish
+        for executor in executors {
+            let _ = executor.await;
```

Comment:
How about awaiting for the multiple futures at the same time using `join`? https://rust-lang.github.io/async-book/06_multiple_futures/02_join.html The for loop makes it not exactly parallel.